internal/webhook/v1beta1/database_webhook.go

```diff
@@ -0,0 +1,167 @@
+)
+
+// nolint:unused
+// log is for logging in this package.
+var databaselog = logf.Log.WithName("database-resource")
+
+const (
+	DEFAULT_TEMPLATE_VALUE = "{{ .Protocol }}://{{ .Username }}:{{ .Password }}@{{ .Hostname }}:{{ .Port }}/{{ .Database }}"
+	DEFAULT_TEMPLATE_NAME  = "CONNECTION_STRING"
+)
+
+// SetupDatabaseWebhookWithManager registers the webhook for Database in the manager.
+func SetupDatabaseWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).For(&kindarocksv1beta1.Database{}).
+		WithValidator(&DatabaseCustomValidator{}).
+		WithDefaulter(&DatabaseCustomDefaulter{}).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/mutate-kinda-rocks-v1beta1-database,mutating=true,failurePolicy=fail,sideEffects=None,groups=kinda.rocks,resources=databases,verbs=create;update,versions=v1beta1,name=mdatabase-v1beta1.kb.io,admissionReviewVersions=v1
+
+type DatabaseCustomDefaulter struct{}
+
+var _ webhook.CustomDefaulter = &DatabaseCustomDefaulter{}
+
+func (d *DatabaseCustomDefaulter) Default(_ context.Context, obj runtime.Object) error {
+	database, ok := obj.(*kindarocksv1beta1.Database)
+
+	if !ok {
+		return fmt.Errorf("expected an Database object but got %T", obj)
+	}
+
+	databaselog.Info("Defaulting for Database", "name", database.GetName())
+
+	if len(database.Spec.SecretsTemplates) == 0 && len(database.Spec.Credentials.Templates) == 0 {
+		database.Spec.Credentials = kindarocksv1beta1.Credentials{
+			Templates: kindarocksv1beta1.Templates{
+				&kindarocksv1beta1.Template{
+					Name:     DEFAULT_TEMPLATE_NAME,
+					Template: DEFAULT_TEMPLATE_VALUE,
+					Secret:   true,
+				},
+			},
+		}
+	}
+
+	return nil
+}
+
+// +kubebuilder:webhook:path=/validate-kinda-rocks-v1beta1-database,mutating=false,failurePolicy=fail,sideEffects=None,groups=kinda.rocks,resources=databases,verbs=create;update,versions=v1beta1,name=vdatabase-v1beta1.kb.io,admissionReviewVersions=v1
+
+type DatabaseCustomValidator struct {
+	// TODO(user): Add more fields as needed for validation
+}
+
+var _ webhook.CustomValidator = &DatabaseCustomValidator{}
+
+// ValidateCreate implements webhook.CustomValidator so a webhook will be registered for the type Database.
+func (v *DatabaseCustomValidator) ValidateCreate(_ context.Context, obj runtime.Object) (admission.Warnings, error) {
+	database, ok := obj.(*kindarocksv1beta1.Database)
+	if !ok {
+		return nil, fmt.Errorf("expected a Database object but got %T", obj)
+	}
+	databaselog.Info("Validation for Database upon creation", "name", database.GetName())
+
+	// TODO(user): fill in your validation logic upon object creation.
+	if database.Spec.SecretsTemplates != nil && database.Spec.Credentials.Templates != nil {
+		return nil, errors.New("using both: secretsTemplates and templates, is not allowed")
+	}
+
+	if database.Spec.SecretsTemplates != nil {
+		databaselog.Info("secretsTemplates are deprecated, it will be removed in the next API version. Please, consider switching to templates")
```

Comment:
Would be more handy to return this also as `return admission.Warnings` for better visibility, or?